jdcoefct.c

```diff
@@ -584,18 +590,20 @@ decompress_smooth_data(j_decompress_ptr cinfo, JSAMPIMAGE output_buf)
           DC20 = (int)next_block_row[2][0];
           DC25 = (int)next_next_block_row[2][0];
         }
-        /* Compute coefficient estimates with a Gaussian-like kernel,
+        /* Compute coefficient estimates with some smoothing. Here the
+         * coefficients come from smoothing with Gaussian like kernel,
          * keeping the averages of the DC values.
          * An estimate is applied only if coefficient is still zero,
          * and is not known to be fully accurate.
          */
         /* AC01 */
         if ((Al = coef_bits[1]) != 0 && workspace[1] == 0) {
-          num = Q00 *
+          num = Q00 * (change_dc ?
                 (-DC01 - DC02 + DC04 + DC05 - 3 * DC06 + 13 * DC07 -
                  13 * DC09 + 3 * DC10 - 3 * DC11 + 38 * DC12 - 38 * DC14 +
                  3 * DC15 - 3 * DC16 + 13 * DC17 - 13 * DC19 + 3 * DC20 -
-                 DC21 - DC22 + DC24 + DC25);
+                 DC21 - DC22 + DC24 + DC25) :
+                 (-7*DC11 + 50*DC12 - 50*DC14 + 7*DC15));
```

Comment:
Sure! There are two different set of numbers:
 1) The last numbers in the question mark operators are a straight forward generalization of the interpolation done in K.8 (appendix K, section 8 in the jpeg standard) to a 5x5 window (they use a 3x3 window here). Basically we assume the the the image is a polynomial of the form A1*x^4*y^4+A2*x^4*y^3+A3*x^2*y^2+... +A23*x + A24*y + A25, then solve a system to get the 25 As and then transfer that calulation into DCT space to get the coefficients. This approach, like 3x3-window K.8 leaves the DC at the cost of looking like less smooth (5x5 looks a little better than 3x3), but more pixelated due to overshoot artifacts.
2) The numbers for the case dc_change == TRUE are calculated differently. Here we apply some smoothing first. The type of smoothing makes sure that on average the DC does not change. (This is why we change the DC below!) 
The smoothing applied is similar to smoohting directly with a Gaussian kernel, but tuned such that is looks appealing and non-disturbing on images. The exact algorithm for the smoothing can be found here: https://gitlab.com/wg1/jpeg-xl/-/blob/master/tools/upscaling_coefficients/generate_upscaling_coefficients.py .
This colab operatores entirely in pixel space so in order to get the coefficients here one has run that code with upscaling_factor=8, (kernel_size=5 is default) and then apply the jpeg-DCT transform to obain a kernel in DCT space. Then we apply the same method as in K.8 turning floats into integers. One subtlety here is that due to rounding the sum of those integers might not be zero, which we definitely want to have, since is would otherwise lead to visible artifacts. We solve that problem by optimizing for a rounding that has as an additional requirement that the sum is zero. 

Please let me know if you need more info on that!